### PR TITLE
Revamp failure testing with query assertions

### PR DESCRIPTION
**Pull request by findepi**
Before the introduction of query assertions (the `QueryAssert` class),
`assertQueryFails` was used to verify that a query fails with expected
error message AND the failure exception type is `TrinoException`.

With query assertions, the typical pattern for testing query failures
became:

    assertThatThrownBy(() -> query(...))
        ....

This unfortunately does not verify that the exception type is the
`TrinoException`. The existing alternative:

    assertTrinoExceptionThrownBy(() -> query(...))
        ...

is rarely used, perhaps because it's longer and the
established shorter pattern is followed.

Both above methods share the disadvantage of basing on the `QueryAssert`
construction side-effect of evaluating the query eagerly.

This commit revamps `QueryAssert` to make failures first class citizen
from `QueryAssert` perspective, on par with query results. Expected
failures are verified to be `TrinoException` type. The result is more
fluent API, as both success and failures assertions are available on the
same query assert object, and thus can be auto-completed by an IDE.

From code updating perspective, most of the code updates fall into two
categories:

* failure testing. This functionality is available from query assert via
  `.failure()` accessor. Code was updated with regular expression
  - search for: `(assertThatThrownBy|assertTrinoExceptionThrownBy)\(\(\) -> (assertions\.)?query\(`
  - replace with: `assertThat($2query(`
  - add `.failure()` to access assertions on the failure exception.
    (use `nonTrinoExceptionFailure()` for cases where query incorrectly
        fails without a proper `TrinoException`)

* detailed inspection of `MaterializedResult` objects. This lower-level
  functionality is available from query assert via `.result()`
  accessor.

### CODE REVIEW

**Comment by findepi** [comment]
cc @trinodb/maintainers

**Comment by findepi** [comment]
currently based on https://github.com/trinodb/trino/pull/20506

File: core/trino-main/src/test/java/io/trino/sql/query/TestAggregation.java [hunk 1]
```diff
@@ -38,22 +37,22 @@ public void teardown()
     @Test
     public void testQuantifiedComparison()
     {
-        assertThatThrownBy(() -> assertions.query("SELECT v > ALL (VALUES 1) FROM (VALUES (1, 1), (1, 2)) t(k, v) GROUP BY k"))
-                .hasMessageContaining("must be an aggregate expression or appear in GROUP BY clause");
+        assertThat(assertions.query("SELECT v > ALL (VALUES 1) FROM (VALUES (1, 1), (1, 2)) t(k, v) GROUP BY k"))
```
Comment (wendigo) [inline core/trino-main/src/test/java/io/trino/sql/query/TestAggregation.java:40]: I like the new API :)

**Review by wendigo — CHANGES_REQUESTED** [review]
@findepi Can we extract changes to the `QueryAssertions` in the last commit to the separate commit as you are mixing mechanical (regex) changes and manual ones in a single commit?

**Comment by wendigo** [comment]
There are related failures

**Comment by findepi** [comment]
> Can we extract changes to the `QueryAssertions` in the last commit to the separate commit as you are mixing mechanical (regex) changes and manual ones in a single commit?

The changes are tightly related, I don't see a way to split them.

File: core/trino-main/src/test/java/io/trino/sql/query/TestAggregation.java (hunk 1 above)
Comment (findepi) [reply core/trino-main/src/test/java/io/trino/sql/query/TestAggregation.java:40]: thank you!

**Review by wendigo — APPROVED** [review]
LGTM % failures

**Comment by findepi** [comment]
The new API requires that failures are a TrinoException (and has an scape hatch for cases where they are not).
When migrating from to the new API, i opted to expected TrinoException everywhere and this is the cause of all the test failures noticed so far. In so many tests we were testing for a failure, but the actual exception is wrong.

No plan to fix these in this PR, but they will now be easy to spot (any usages of `nonTrinoExceptionFailure()`).

**Review by losipiuk — COMMENTED** [review]
looks nice

**Comment by wendigo** [comment]
BigQuery related: 
```
TestBigQueryAvroConnectorTest>BaseBigQueryConnectorTest.testDateYearOfEraPredicate:607 Expected TrinoException or wrapper, but got: io.trino.testing.QueryFailedException io.trino.testing.QueryFailedException: io.grpc.StatusRuntimeException: INVALID_ARGUMENT: request failed: Query error: Could not cast literal "-1996-09-14" to type DATE at [1:372]
Error:    TestBigQueryAvroConnectorTest>BaseBigQueryConnectorTest.testMissingWildcardTable:799 Expected TrinoException or wrapper, but got: io.trino.testing.QueryFailedException io.trino.testing.QueryFailedException: sep-bq-cicd:test.test_missing_wildcard_table_* does not match any table.
Error:    TestBigQueryAvroConnectorTest>BaseBigQueryConnectorTest.testNativeQuerySelectUnsupportedType:938 Expected TrinoException or wrapper, but got: io.trino.testing.QueryFailedException io.trino.testing.QueryFailedException: Unsupported type: BIGNUMERIC
```

**Pull request merged**